### PR TITLE
[SPARK-9681] [ML] Support R feature interactions in RFormula

### PR DESCRIPTION
**Pull request by ericl**
This integrates the Interaction feature transformer with SparkR R formula support (i.e. support `:`).

To generate reasonable ML attribute names for feature interactions, it was necessary to add the ability to read attribute the original attribute names back from `StructField`, and also to specify custom group prefixes in `VectorAssembler`. This also has the side-benefit of cleaning up the double-underscores in the attributes generated for non-interaction terms.

@mengxr

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 1]
```diff
@@ -31,20 +32,28 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
    * of the special '.' term. Duplicate terms will be removed during resolution.
    */
   def resolve(schema: StructType): ResolvedRFormula = {
-    var includedTerms = Seq[String]()
+    val dotTerms = expandDot(schema)
+    var includedTerms = Seq[Seq[String]]()
     terms.foreach {
+      case term: ColumnRef =>
+        includedTerms :+= Seq(term.value)
+      case ColumnInteraction(terms) =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: `terms` shadows the class member `terms`. `terms` -> `cols`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 2]
```diff
@@ -31,20 +32,28 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
...
-      case ColumnRef(value) =>
-        includedTerms :+= value
+        includedTerms ++= dotTerms.map(Seq(_))
       case Deletion(term: Term) =>
         term match {
-          case ColumnRef(value) =>
-            includedTerms = includedTerms.filter(_ != value)
+          case inner: ColumnRef =>
+            includedTerms = includedTerms.filter(_ != Seq(inner.value))
+          case ColumnInteraction(terms) =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: ditto

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 3]
```diff
@@ -31,20 +32,28 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
...
           case Dot =>
             // e.g. "- .", which removes all first-order terms
-            val fromSchema = simpleTypes(schema)
-            includedTerms = includedTerms.filter(fromSchema.contains(_))
+            includedTerms = includedTerms.filter {
+              case Seq(t) => !dotTerms.contains(t)
+              case _ => true
+            }
           case _: Deletion =>
             assert(false, "Deletion terms cannot be nested")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: * not part of this PR: `throw new RuntimeException(...)`
* also not part of this PR: Shall we move `hasIntercept` to `ResolvedRFormula`? It is a little strange to have two places to store the fully parsed formula.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 4]
```diff
@@ -67,19 +76,52 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
...
+  private def expandInteraction(
+      schema: StructType, terms: Seq[InteractionComponent]): Seq[Seq[String]] = {
+    if (terms.isEmpty) {
+      return Seq(Nil)
+    }
+
+    val rest = expandInteraction(schema, terms.tail)
+    val validInteractions = (terms.head match {
+      case Dot =>
+        expandDot(schema).filter(_ != label.value).flatMap { t =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: remove `.filter(_ != label.value)`

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 5]
```diff
@@ -87,11 +129,17 @@ private[ml] case class ResolvedRFormula(label: String, terms: Seq[String])
...
+/** A term that may be part of an interaction, e.g. 'x' in 'x:y' */
+private[ml] sealed trait InteractionComponent extends Term
+
 /* R formula reference to all available columns, e.g. "." in a formula */
-private[ml] case object Dot extends Term
+private[ml] case object Dot extends InteractionComponent
 
 /* R formula reference to a column, e.g. "+ Species" in a formula */
-private[ml] case class ColumnRef(value: String) extends Term
+private[ml] case class ColumnRef(value: String) extends InteractionComponent
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: This makes the implementation easier but code harder to understand, because a `ColRef` is not an interaction.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 6]
```diff
@@ -109,7 +157,15 @@ private[ml] object RFormulaParser extends RegexParsers {
   def columnRef: Parser[ColumnRef] =
     "([a-zA-Z]|\\.[a-zA-Z_])[a-zA-Z0-9._]*".r ^^ { case a => ColumnRef(a) }
 
-  def term: Parser[Term] = intercept | columnRef | "\\.".r ^^ { case _ => Dot }
+  def dot: Parser[InteractionComponent] = "\\.".r ^^ { case _ => Dot }
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: minor: Those could be `val` instead of `def`, right?

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 7]
```diff
@@ -109,7 +157,15 @@ private[ml] object RFormulaParser extends RegexParsers {
...
+
+  def interaction: Parser[List[InteractionComponent]] = repsep(columnRef | dot, ":")
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: If we want to separate `columnRef` from `interaction`, we should use `rep1sep` here and update `term`. This might add several lines of code, but make the code easier to understand.

**Comment by mengxr** [comment]
I only checked `RFormulaParser.scala`. Need more time to go through the rest.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 1 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 2 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 3 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done (though, it's not really needed in ResolvedRFormula and a bit awkward to extract for use in SparkRWrappers)

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 4 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 5 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: I updated the parser so that Interaction doesn't entirely subsume ColRef, let me know what you think.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 6 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala (hunk 7 above)
Comment (ericl) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: Done

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala [hunk 8]
```diff
@@ -47,6 +48,28 @@ class VectorAssembler(override val uid: String)
...
 
+  /**
+   * By default, the attribute names of vector components will be `groupName + '_' + attrName`.
+   * This parameter allows the overriding of the group prefix per input column vector.
+   *
+   * @group param Mapping of input vector names to group prefixes. If not specified, the group
+   *              prefix for an input vector column will default to `groupName + '_'`.
+   * @param groupPrefixes
+   */
+  final val groupPrefixes: Param[Map[String, String]] = new Param(
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala]: This would be a really advanced parameter for users. For this PR, is it possible to use existing attribute name generation in `VectorAssembler` and rename the generated attributes after? We could assign a unique prefix to new columns and then remove this prefix from all attribute names.

Otherwise, there are couple issues with this parameter:
1. It is not Java/Python friendly.
2. It is quite hard to understand.
I don't think those issues should be addressed in this PR.

File: mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala [hunk 9]
```diff
@@ -79,4 +87,79 @@ class RFormulaParserSuite extends SparkFunSuite {
...
+    checkParse("y ~ a : b : c", "y", Seq("a:b:c"))
+    checkParse("y ~ q + a:b:c + b:c + c:d + z", "y", Seq("q", "a:b:c", "b:c", "c:d", "z"))
+  }
+
+  test("parse basic interactions with dot") {
+    val schema = (new StructType)
+      .add("a", "int", true)
+      .add("b", "long", false)
+      .add("c", "string", true)
+    checkParse("y ~ .:x", "y", Seq("a:x", "b:x", "c:x"), schema)
```
Comment (mengxr) [inline mllib/src/test/scala/org/apache/spark/ml/feature/RFormulaParserSuite.scala]: This is a little confusing because `x` and `y` do not appear in the schema.

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 10]
```diff
@@ -31,27 +32,35 @@ private[ml] case class ParsedRFormula(label: ColumnRef, terms: Seq[Term]) {
    * of the special '.' term. Duplicate terms will be removed during resolution.
    */
   def resolve(schema: StructType): ResolvedRFormula = {
-    var includedTerms = Seq[String]()
+    val dotTerms = expandDot(schema)
+    var includedTerms = Seq[Seq[String]]()
     terms.foreach {
+      case term: ColumnRef =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala]: minor: `term` -> `col`

File: mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala [hunk 11]
```diff
@@ -70,10 +93,10 @@ class VectorAssembler(override val uid: String)
           val group = AttributeGroup.fromStructField(field)
           if (group.attributes.isDefined) {
             // If attributes are defined, copy them with updated names.
+            val prefix = $(groupPrefixes).get(c).getOrElse(c + "_")
             group.attributes.get.map { attr =>
               if (attr.name.isDefined) {
-                // TODO: Define a rigorous naming scheme.
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/feature/VectorAssembler.scala]: Please keep this TODO.

**Comment by mengxr** [comment]
@ericl I think the only issue left is the `groupPrefixes` param in `VectorAssembler`. It would be nice to keep the feature name handling under the hood.

**Comment by ericl** [comment]
comment should be addressed

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!
